pkg/controller/baremetalhost/host_state_machine.go

```diff
@@ -135,21 +130,43 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
-func (hsm *hostStateMachine) shouldInitiateRegister(info *reconcileInfo) bool {
-	changeState := false
-	if hsm.Host.DeletionTimestamp.IsZero() {
-		switch hsm.NextState {
-		default:
-			changeState = !hsm.Host.Status.GoodCredentials.Match(*info.bmcCredsSecret)
-		case metal3v1alpha1.StateNone, metal3v1alpha1.StateUnmanaged:
-		case metal3v1alpha1.StateRegistering, metal3v1alpha1.StateRegistrationError:
-		case metal3v1alpha1.StateDeleting:
+func (hsm *hostStateMachine) ensureRegistered(info *reconcileInfo) (result actionResult) {
+	switch hsm.NextState {
+	case metal3v1alpha1.StateNone, metal3v1alpha1.StateUnmanaged:
+		// We haven't yet reached the Registration state, so don't attempt
+		// to register the Host.
+		return
+	}
+	if !hsm.Host.DeletionTimestamp.IsZero() {
+		// BUG(zaneb) We currently don't attempt to re-register the Host
+		// if we find it missing once a delete has been requested (in
+		// part because we are also willing to pass empty credentials
+		// after this point), but this means that we may not be able to
+		// deprovision a Host if the Ironic database pod is rescheduled
+		// during deprovisioning, or if the credentials need to be
+		// modified.
+		return
+	}
+
+	if hsm.Host.Status.GoodCredentials.Match(*info.bmcCredsSecret) {
+		// Credentials are unchanged since we verified them.
+		return
+	}
+
+	recordStateBegin(hsm.Host, metal3v1alpha1.StateRegistering, metav1.Now())
+	if hsm.Host.Status.ErrorType == metal3v1alpha1.RegistrationError {
+		if hsm.Host.Status.TriedCredentials.Match(*info.bmcCredsSecret) {
```

Comment:
Maybe we don't need this test any more? Or do we need to wait for #610?

Reply:
Yeah, I suspect the existence of this code is a mistake - registration could have failed for a reason other than the creds being wrong, so we probably shouldn't force the user to update them before we look again. #610 provides an exponential backoff on failure, which is what we really want.
This is bug-for-bug compatible with how it works now, so any changes should go in a separate PR.

Reply:
What will happen incase we had a BMC connectivity issue for a brief moment, and the BMH went to `RegistrationError`, and then after a while the connectivity is reinstated. The workflow you are proposing here won't try to register again since the change in scenario was not related to BMC credentials but related to BMC connectivity. Am I understanding it right?

Reply:
Yes, but with a caveat: it likely wouldn't have gone into RegistrationError before either - it could easily have been ProvisioningError or PowerManagementError, because we don't know the cause of the error, only what we were trying to do when it happened.

In those scenarios it's fine anyway, because we don't need to re-register (i.e. send new credentials to Ironic) to reconnect; a retry is enough. The theory behind this code is probably that the only circumstances we need to re-register is if the password is changed on the BMC, and in that case it won't work until the creds are updated in k8s anyway so there's no point retrying until they are.

The problem with that logic is that if we do change the credentials and connectivity is interrupted *while we are updating the credentials* then we erroneously assume it was because the credentials are bad and don't try again. (The same would presumably apply if you change them in the 'wrong' order - in the k8s Secret first and later on the BMC.) If that proves to be the case, it's a (pre-existing) bug.

Reply:
> In those scenarios it's fine anyway, because we don't need to re-register (i.e. send new credentials to Ironic) to reconnect; a retry is enough. The theory behind this code is probably that the only circumstances we need to re-register is if the password is changed on the BMC, and in that case it won't work until the creds are updated in k8s anyway so there's no point retrying until they are.
> 

Does this mean that in case of BMC connectivity issue, no extra logic needs to be added to host state machine and retry is already implemented? I am interested in this specific case where only bmc connectivity is lost and nothing wrong happened with the credentials. I would like to know your opinion on that specific case, and what needs to be done to get BMH out of `RegistrationError` @zaneb ? Thanks!

Reply:
> Does this mean that in case of BMC connectivity issue, no extra logic needs to be added to host state machine and retry is already implemented?

Yes, as of #610 retry is implemented.
(Prior to that we in theory stopped reconciling after a registration error, and waited for some change to the Host or Secret. In practice I suspect that in many cases writes to the status would have triggered an immediate re-reconcile and resultant retry anyway.)